### PR TITLE
Support runAsNonRoot true for recent versions of EPR

### LINKED ISSUE #8968: Enable `RunAsNonRoot: true` for Package Registry
### Background

PR #8800 adds support for deploying package-registry in ECK, but currently does not enforce `RunAsNonRoot` in the security context because the package-registry images still run as root, with UID 0.

This PR https://github.com/elastic/package-registry/pull/1503 to transition the container to run as user `UID 1000`, similar to other Elastic stack components.

### Goal

Once the changes from elastic/package-registry#1503 are merged and released in `docker.elastic.co/package-registry/distribution` images, we should update ECK to enforce `RunAsNonRoot: true` for EPR instances.

### Additional Context

- Related PR: #8800
- This change will improve the security posture of package-registry deployments in ECK
- We should ensure backward compatibility if ECK needs to support registry versions both with and without this change

### Questions to resolve

- Which package-registry versions will include the user 1000 change?

### PR DESCRIPTION
**Pull request by pkoutsovasilis**
## Summary

This PR adds support for `runAsNonRoot: true` in the security context for Elastic Package Registry (EPR) pods on versions that support running as non-root.

Following [elastic/package-registry#1503](https://github.com/elastic/package-registry/pull/1503), newer versions of the package registry can run as a non-root user.

 This change enables the `runAsNonRoot` security context setting for the following versions:

- **9.3.0+**
- **9.2.4+**
- **9.1.10+**
- **8.19.10+**

For older versions, `runAsNonRoot` is left unset (nil) to maintain backward compatibility.

## Changes

- Added version detection logic in `pkg/controller/packageregistry/pod.go` to conditionally set `runAsNonRoot: true` based on the EPR version
- Added comprehensive unit tests in `pkg/controller/packageregistry/pod_test.go` covering all version boundary conditions

## Related Issue

Closes https://github.com/elastic/cloud-on-k8s/issues/8968

### CODE REVIEW

**Comment by prodsecmachine** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/elastic-cloud-on-k8s/pr-checks/c3ee512b-cd45-4630-bcb3-83658abd9f26?refs=fad745fcac78318c51f38a50636e4e22ae206cf9&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/elastic-cloud-on-k8s/pr-checks/c3ee512b-cd45-4630-bcb3-83658abd9f26/license?refs=fad745fcac78318c51f38a50636e4e22ae206cf9&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: pkg/controller/packageregistry/pod.go [hunk 1]
```diff
@@ -91,6 +91,30 @@ func newPodSpec(epr eprv1alpha1.PackageRegistry, configHash string, meta metadat
...
+	case v.Major == 9 && v.Minor == 1 && v.Patch >= 10:
+		// version 9.1.10+
+		runAsNonRoot = ptr.To(true)
+	case v.Major == 8 && v.Minor >= 20:
+		// version 8.20+
+		runAsNonRoot = ptr.To(true)
+	case v.Major == 8 && v.Minor == 19 && v.Patch >= 10:
+		// version 8.19.10+
+		runAsNonRoot = ptr.To(true)
+	}
```
Comment (barkbay) [inline pkg/controller/packageregistry/pod.go]: LGTM, but I must confess I find `version.From()` more readable, I feel it expresses the intent more clearly. Also, this might deserve a dedicated function:

```golang
// supportsRunAsNonRoot returns true if the given version supports running as non-root.
// See https://github.com/elastic/package-registry/pull/1503
func supportsRunAsNonRoot(v version.Version) bool {
	// Strip pre-release suffix (e.g., -SNAPSHOT) for comparison since
	// SNAPSHOT builds of a version include the feature if the release does.
	v = version.WithoutPre(v)

	// 9.3.0+ (including 9.4+, 10+, etc.)
	if v.GTE(version.From(9, 3, 0)) {
		return true
	}
	// 9.2.4+ (backport to 9.2.x line)
	if v.GTE(version.From(9, 2, 4)) && v.LT(version.From(9, 3, 0)) {
		return true
	}
	// 9.1.10+ (backport to 9.1.x line)
	if v.GTE(version.From(9, 1, 10)) && v.LT(version.From(9, 2, 0)) {
		return true
	}
	// 8.19.10+ (backport to 8.19.x and 8.20+ lines)
	if v.GTE(version.From(8, 19, 10)) && v.LT(version.From(9, 0, 0)) {
		return true
	}
	return false
}
```

This avoid `8.20+` case for example (which is not in the comment btw).

**Review by barkbay — APPROVED** [review]

File: pkg/controller/packageregistry/pod_test.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		{
+			name:               "version 8.18.0 should have runAsNonRoot set to nil",
+			epr:                getEprWithVersion("8.18.0"),
+			expectRunAsNonRoot: false,
+		},
+		{
+			name:               "version 8.20.0 should have runAsNonRoot set to true",
+			epr:                getEprWithVersion("8.20.0"),
+			expectRunAsNonRoot: true,
+		},
```
Comment (barkbay) [inline pkg/controller/packageregistry/pod_test.go:121]:
```suggestion
		},
		// Pre-release versions (SNAPSHOT, rc, etc.) at or above the threshold SHOULD have the feature
		// because SNAPSHOT/pre-release builds include the merged feature code
		{
			name:               "version 8.19.10-SNAPSHOT should have runAsNonRoot set to true",
			epr:                getEprWithVersion("8.19.10-SNAPSHOT"),
			expectRunAsNonRoot: true,
		},
		{
			name:               "version 9.3.0-SNAPSHOT should have runAsNonRoot set to true",
			epr:                getEprWithVersion("9.3.0-SNAPSHOT"),
			expectRunAsNonRoot: true,
		},
		// Pre-release versions BELOW the threshold should NOT have the feature
		{
			name:               "version 8.19.9-SNAPSHOT should have runAsNonRoot set to nil",
			epr:                getEprWithVersion("8.19.9-SNAPSHOT"),
			expectRunAsNonRoot: false,
		},
		{
			name:               "version 9.2.3-SNAPSHOT should have runAsNonRoot set to nil",
			epr:                getEprWithVersion("9.2.3-SNAPSHOT"),
			expectRunAsNonRoot: false,
		},
```

File: pkg/controller/packageregistry/pod.go (hunk 1 above)
Comment (pkoutsovasilis) [reply pkg/controller/packageregistry/pod.go]: Confession time, honestly, I had the same debate myself but eventually I gave up and went with plain checks. But since you brought it up, it makes total sense and with the dedicated function this is almost beautiful (I would still prefer not have to do a check here and this be always true 😆). Changed in fad745fcac78318c51f38a50636e4e22ae206cf9

**Comment by pkoutsovasilis** [comment]
buildkite test this -f p=kind,s=9.2.4,t=TestPackageRegistry.*
